### PR TITLE
MLflow Gateway Example

### PR DESCRIPTION
**Pull request by harupy**
# See https://github.com/harupy/langchain/tree/gateway-integration/examples for how to run the examples.

### CODE REVIEW

File: langchain/llms/mlflow.py [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+    temperature: float = 0.3
+
+    def _call(
+        self,
+        prompt: str,
+        stop: List[str] | None = None,
+        run_manager: CallbackManagerForLLMRun | None = None,
+        **kwargs: Any
+    ) -> str:
+        return "mlflow-gateway"
```
Comment (harupy) [inline langchain/llms/mlflow.py]:
```suggestion
        return mlflow.gateway.query(...)
```

This part would look like this?

File: examples/completions.py
Comment (harupy) [inline examples/completions.py:1]: This script runs fine and prints out `'mlflow-gateway'`

File: langchain/llms/mlflow.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+        **kwargs: Any,
+    ) -> str:
+        resp = mlflow.gateway.query(
+            self.route,
+            data={
+                "temperature": self.temperature,
+                "prompt": prompt,
+            },
+        )
+        return resp["candidates"][0]["text"]
```
Comment (harupy) [inline langchain/llms/mlflow.py]: Didn't notcie `kwargs`. I think we can pass this to `query`.

File: langchain/llms/mlflow.py [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+    def __init__(self, **kwargs: Any):
+        import mlflow
+
+        super().__init__(**kwargs)
+        mlflow.gateway.set_gateway_uri(self.gateway_uri)
+
+    def _call(
+        self,
+        prompt: str,
+        stop: List[str] | None = None,
```
Comment (harupy) [inline langchain/llms/mlflow.py:51]: `stop` needs to be passed to `query`

**Comment by harupy** [comment]
```python
import mlflow
from langchain import LLMChain, PromptTemplate
from langchain.llms import MlflowGateway


gateway_llm = MlflowGateway(
    gateway_uri="http://127.0.0.1:5000",
    route="completions",
    temperature=0.3,
)

llm_chain = LLMChain(
    llm=gateway_llm,
    prompt=PromptTemplate(input_variables=["adjective"], template="I am a {adjective}"),
)
result = llm_chain.run(adjective="professional")
print(result)


with mlflow.start_run():
    info = mlflow.langchain.log_model(llm_chain, "model")
    print(mlflow.langchain.load_model(info.model_uri))

```

threw:

```
2023/06/15 16:45:55 WARNING mlflow: MLflow does not guarantee support for LLMs outside of HuggingFaceHub and OpenAI, found MlflowGateway
2023/06/15 16:45:57 WARNING mlflow.utils.environment: Encountered an unexpected error while inferring pip requirements (model URI: /tmp/tmp70wrmp9o/model, flavor: langchain), fall back to return ['langchain==0.0.200']. Set logging level to DEBUG to see the full traceback.
Traceback (most recent call last):
  File "/home/haru/Desktop/repositories/langchain/examples/completions.py", line 22, in <module>
    print(mlflow.langchain.load_model(info.model_uri))
  File "/home/haru/Desktop/repositories/langchain/.venv/lib/python3.10/site-packages/mlflow/langchain/__init__.py", line 516, in load_model
    return _load_model_from_local_fs(local_model_path)
  File "/home/haru/Desktop/repositories/langchain/.venv/lib/python3.10/site-packages/mlflow/langchain/__init__.py", line 491, in _load_model_from_local_fs
    return _load_model(lc_model_path, agent_model_path, tools_model_path, agent_primitive_path)
  File "/home/haru/Desktop/repositories/langchain/.venv/lib/python3.10/site-packages/mlflow/langchain/__init__.py", line 402, in _load_model
    model = load_chain(path)
  File "/home/haru/Desktop/repositories/langchain/langchain/chains/loading.py", line 508, in load_chain
    return _load_chain_from_file(path, **kwargs)
  File "/home/haru/Desktop/repositories/langchain/langchain/chains/loading.py", line 524, in _load_chain_from_file
    config = yaml.safe_load(f)
  File "/home/haru/Desktop/repositories/langchain/.venv/lib/python3.10/site-packages/yaml/__init__.py", line 125, in safe_load
    return load(stream, SafeLoader)
  File "/home/haru/Desktop/repositories/langchain/.venv/lib/python3.10/site-packages/yaml/__init__.py", line 81, in load
    return loader.get_single_data()
  File "/home/haru/Desktop/repositories/langchain/.venv/lib/python3.10/site-packages/yaml/constructor.py", line 51, in get_single_data
    return self.construct_document(node)
  File "/home/haru/Desktop/repositories/langchain/.venv/lib/python3.10/site-packages/yaml/constructor.py", line 60, in construct_document
    for dummy in generator:
  File "/home/haru/Desktop/repositories/langchain/.venv/lib/python3.10/site-packages/yaml/constructor.py", line 413, in construct_yaml_map
    value = self.construct_mapping(node)
  File "/home/haru/Desktop/repositories/langchain/.venv/lib/python3.10/site-packages/yaml/constructor.py", line 218, in construct_mapping
    return super().construct_mapping(node, deep=deep)
  File "/home/haru/Desktop/repositories/langchain/.venv/lib/python3.10/site-packages/yaml/constructor.py", line 143, in construct_mapping
    value = self.construct_object(value_node, deep=deep)
  File "/home/haru/Desktop/repositories/langchain/.venv/lib/python3.10/site-packages/yaml/constructor.py", line 100, in construct_object
    data = constructor(self, node)
  File "/home/haru/Desktop/repositories/langchain/.venv/lib/python3.10/site-packages/yaml/constructor.py", line 427, in construct_undefined
    raise ConstructorError(None, None,
yaml.constructor.ConstructorError: could not determine a constructor for the tag 'tag:yaml.org,2002:python/object/apply:builtins.getattr'
  in "/home/haru/Desktop/repositories/langchain/mlruns/0/da23b407f4fa4f7286763f7755727c6d/artifacts/model/model.yaml", line 3, column 10
```

File: examples/completions.py
Comment (harupy) [inline examples/completions.py:1]:
```python
import mlflow
from langchain import LLMChain, PromptTemplate
from langchain.llms import MlflowGateway


gateway_llm = MlflowGateway(
    gateway_uri="http://127.0.0.1:5000",
    route="completions",
    temperature=0.3,
)

llm_chain = LLMChain(
    llm=gateway_llm,
    prompt=PromptTemplate(input_variables=["adjective"], template="I am a {adjective}"),
)
result = llm_chain.run(adjective="professional")
print(result)


with mlflow.start_run():
    info = mlflow.langchain.log_model(llm_chain, "model")
    print(mlflow.langchain.load_model(info.model_uri))

```

threw:

```
2023/06/15 16:45:55 WARNING mlflow: MLflow does not guarantee support for LLMs outside of HuggingFaceHub and OpenAI, found MlflowGateway
2023/06/15 16:45:57 WARNING mlflow.utils.environment: Encountered an unexpected error while inferring pip requirements (model URI: /tmp/tmp70wrmp9o/model, flavor: langchain), fall back to return ['langchain==0.0.200']. Set logging level to DEBUG to see the full traceback.
Traceback (most recent call last):
  File "/home/haru/Desktop/repositories/langchain/examples/completions.py", line 22, in <module>
    print(mlflow.langchain.load_model(info.model_uri))
  File "/home/haru/Desktop/repositories/langchain/.venv/lib/python3.10/site-packages/mlflow/langchain/__init__.py", line 516, in load_model
    return _load_model_from_local_fs(local_model_path)
  File "/home/haru/Desktop/repositories/langchain/.venv/lib/python3.10/site-packages/mlflow/langchain/__init__.py", line 491, in _load_model_from_local_fs
    return _load_model(lc_model_path, agent_model_path, tools_model_path, agent_primitive_path)
  File "/home/haru/Desktop/repositories/langchain/.venv/lib/python3.10/site-packages/mlflow/langchain/__init__.py", line 402, in _load_model
    model = load_chain(path)
  File "/home/haru/Desktop/repositories/langchain/langchain/chains/loading.py", line 508, in load_chain
    return _load_chain_from_file(path, **kwargs)
  File "/home/haru/Desktop/repositories/langchain/langchain/chains/loading.py", line 524, in _load_chain_from_file
    config = yaml.safe_load(f)
  File "/home/haru/Desktop/repositories/langchain/.venv/lib/python3.10/site-packages/yaml/__init__.py", line 125, in safe_load
    return load(stream, SafeLoader)
  File "/home/haru/Desktop/repositories/langchain/.venv/lib/python3.10/site-packages/yaml/__init__.py", line 81, in load
    return loader.get_single_data()
  File "/home/haru/Desktop/repositories/langchain/.venv/lib/python3.10/site-packages/yaml/constructor.py", line 51, in get_single_data
    return self.construct_document(node)
  File "/home/haru/Desktop/repositories/langchain/.venv/lib/python3.10/site-packages/yaml/constructor.py", line 60, in construct_document
    for dummy in generator:
  File "/home/haru/Desktop/repositories/langchain/.venv/lib/python3.10/site-packages/yaml/constructor.py", line 413, in construct_yaml_map
    value = self.construct_mapping(node)
  File "/home/haru/Desktop/repositories/langchain/.venv/lib/python3.10/site-packages/yaml/constructor.py", line 218, in construct_mapping
    return super().construct_mapping(node, deep=deep)
  File "/home/haru/Desktop/repositories/langchain/.venv/lib/python3.10/site-packages/yaml/constructor.py", line 143, in construct_mapping
    value = self.construct_object(value_node, deep=deep)
  File "/home/haru/Desktop/repositories/langchain/.venv/lib/python3.10/site-packages/yaml/constructor.py", line 100, in construct_object
    data = constructor(self, node)
  File "/home/haru/Desktop/repositories/langchain/.venv/lib/python3.10/site-packages/yaml/constructor.py", line 427, in construct_undefined
    raise ConstructorError(None, None,
yaml.constructor.ConstructorError: could not determine a constructor for the tag 'tag:yaml.org,2002:python/object/apply:builtins.getattr'
  in "/home/haru/Desktop/repositories/langchain/mlruns/0/da23b407f4fa4f7286763f7755727c6d/artifacts/model/model.yaml", line 3, column 10
```

Comment (harupy) [reply examples/completions.py:1]: `model.yaml` looks like this:

```yaml
_type: llm_chain
llm:
  _type: !!python/object/apply:builtins.getattr
  - !!python/object:langchain.llms.mlflow.MlflowGateway
    __dict__:
      cache: null
      callback_manager: null
      callbacks: null
      gateway_uri: http://127.0.0.1:5000
      lc_kwargs:
        gateway_uri: http://127.0.0.1:5000
        route: completions
        temperature: 0.3
      route: completions
      temperature: 0.3
      verbose: false
    __fields_set__: !!set
      gateway_uri: null
      lc_kwargs: null
      route: null
      temperature: null
    __private_attribute_values__: {}
  - _llm_type
memory: null
output_key: text
prompt:
  _type: prompt
  input_variables:
  - adjective
  output_parser: null
  partial_variables: {}
  template: I am a {adjective}
  template_format: f-string
  validate_template: true
tags: null
verbose: false

```

File: langchain/llms/mlflow.py [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+from typing import Any, List, Dict, Mapping
+from langchain.callbacks.manager import CallbackManagerForLLMRun
+from langchain.llms.base import LLM
+import mlflow.gateway
+
+
+class MlflowGateway(LLM):
+    gateway_uri: str
+    route: str
+    temperature: float = 0.3
```
Comment (dbczumar) [inline langchain/llms/mlflow.py]: We'll need to support additional standard parameters from https://docs.google.com/document/d/1AEDlqSWh4n0j1Ue-llk03aRz7-xBK4P4Zh-Yvo1uhnI/edit like max_tokens, right?

Comment (dbczumar) [reply langchain/llms/mlflow.py]: We should also adjust the default value to be 0

Comment (harupy) [reply langchain/llms/mlflow.py]: I'll add them and fix the default value.

File: langchain/embeddings/mlflow_gateway.py [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+    gateway_uri: str
+    route: str
+
+    def __init__(self, **kwargs: Any):
+        import mlflow
+
+        super().__init__(**kwargs)
+        mlflow.gateway.set_gateway_uri(self.gateway_uri)
+
+    def _query(self, texts: List[str]) -> List[float]:
```
Comment (harupy) [inline langchain/embeddings/mlflow_gateway.py]:
```suggestion
    def _query(self, texts: List[str]) -> List[List[float]]:
```

File: langchain/llms/mlflow.py [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+from langchain.llms.base import LLM
+import mlflow.gateway
+
+
+class MlflowGateway(LLM):
+    gateway_uri: str
+    route: str
+    temperature: float = 0.0
+    stop: List[str] | None = None
+    max_tokens: int = 0
```
Comment (dbczumar) [inline langchain/llms/mlflow.py]: @BenWilson2: I think this should be None (optional).

Comment (harupy) [inline langchain/llms/mlflow.py]:
```suggestion
    max_tokens: Optional[int] = None
```

File: langchain/llms/mlflow.py [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+
+        super().__init__(**kwargs)
+        mlflow.gateway.set_gateway_uri(self.gateway_uri)
+
+    @property
+    def _default_params(self) -> Dict[str, Any]:
+        return {
+            "gateway_uri": self.gateway_uri,
+            "route": self.route,
+            "temperature": self.temperature,
```
Comment (harupy) [inline langchain/llms/mlflow.py]:
```suggestion
            "temperature": self.temperature,
            ...
```

TODO: Add parameters

File: langchain/llms/mlflow.py [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+    temperature: float = 0.0
+    stop: List[str] | None = None
+    max_tokens: Optional[int] = None
+    candidate_count: int = 5
+
+    def __init__(self, **kwargs: Any):
+        import mlflow
+
+        super().__init__(**kwargs)
+        mlflow.gateway.set_gateway_uri(self.gateway_uri)
```
Comment (harupy) [inline langchain/llms/mlflow.py]: Might be better to use the client to avoid side-effects.

File: langchain/llms/mlflow.py [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+    temperature: float = 0.0
+    stop: List[str] | None = None
+    max_tokens: Optional[int] = None
+    candidate_count: int = 5
+
+
+class MlflowGateway(LLM):
+    gateway_uri: str
+    route: str
+    params: Params | None = None
```
Comment (harupy) [inline langchain/llms/mlflow.py]: @dbczumar Added `params` which validates the common params and also accepts extra params.

File: langchain/llms/mlflow.py [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+
+if TYPE_CHECKING:
+    import mlflow.gateway
+
+
+@lru_cache()
+def _get_client(gateway_uri: str) -> mlflow.gateway.MlflowGatewayClient:
+    import mlflow.gateway
+
+    return mlflow.gateway.MlflowGatewayClient(gateway_uri)
```
Comment (harupy) [inline langchain/llms/mlflow.py]: Tried `self.client = mlflow.gateway.MlflowGatewayClient(...)` first, but rejected by pydantic, then switched to this approach.

**Review by dbczumar — APPROVED** [review]
LGTM! Thanks @harupy !!!

File: langchain/embeddings/mlflow_gateway.py [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+        super().__init__(**kwargs)
+        mlflow.gateway.set_gateway_uri(self.gateway_uri)
+
+    def _query(self, texts: List[str]) -> List[List[float]]:
+        embeddings = []
+        for t in chunk(texts, 1000):
+            resp = _get_client(self.gateway_uri).query(
+            self.route,
+            data={"text": t},
+        )
```
Comment (dbczumar) [inline langchain/embeddings/mlflow_gateway.py]: @harupy Can we use a smaller batch size? We're going to need to limit the batch size we support on Databricks to protect the health of the service (avoid OOMs).

@mingyu89 I recall that we discussed a max batch size of 50 initially. Does that still work?

Comment (mingyu89) [reply langchain/embeddings/mlflow_gateway.py]: https://docs.google.com/document/d/19bTRlKFNDxDczBMebv1aj0fe_eOk5jK4wIrPeX7-WeA/edit
Without streaming support, if we use a buffer size of 640k for each response, we'll be able to support batch size up to 32.
I would recommend 20, 25(or 30) to be safer.

Comment (dbczumar) [reply langchain/embeddings/mlflow_gateway.py]: Awesome, okay. Let's stick with 20. Best to be safe; we can scale up later.

File: langchain/embeddings/mlflow_gateway.py [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+
+    def __init__(self, **kwargs: Any):
+        import mlflow
+
+        super().__init__(**kwargs)
+        mlflow.gateway.set_gateway_uri(self.gateway_uri)
+
+    def _query(self, texts: List[str]) -> List[List[float]]:
+        embeddings = []
+        for t in chunk(texts, 1000):
```
Comment (harupy) [inline langchain/embeddings/mlflow_gateway.py]:
```suggestion
        for t in chunk(texts, 20):
```

**Comment by harupy** [comment]
https://github.com/hwchase17/langchain/pull/7113

**Pull request closed without merging**